install_config/upgrading/migrating_etcd.adoc

```diff
@@ -0,0 +1,274 @@
++
+When complete, the migration responds with the following message if successful:
++
+----
+finished transforming keys
+----
++
+If there is no v2 data, it responds with:
++
+----
+no v2 keys to migrate
+----
+
+.. If the `etcdctl migrate` command is run without the `--no-ttl` option, TLL keys
+are migrated as well. Given that the TTL keys in v2 data are replaced with leases in
+v3 data, you must attach leases to all migrated TTL keys by running the following (with
+*etcd* daemon running):
+
+... Create a new cluster on the first host:
++
+----
+# echo "ETCD_FORCE_NEW_CLUSTER=true" >> /etc/etcd/etcd.conf
+# systemctl start etcd
+# sed -i '/ETCD_FORCE_NEW_CLUSTER=true/d' /etc/etcd/etcd.conf
+# systemctl restart etcd
+----
+
+... Scale up additional etcd hosts by following the
+xref:/backup_restore.adoc#adding-addtl-etcd-members[Adding Additional etcd
+Members] documentation.
+
+... After your etcd cluster is back online with all members, re-introduce the TTL
+information by running the following on the first master:
++
+----
+$ oadm migrate etcd-ttl <certificate_details> <endpoint> \
+    --ttl-keys-prefix '/kubernetes.io/events' \
+    --lease-duration 1h
+
+$ oadm migrate etcd-ttl <certificate_details> <endpoint> \
+    --ttl-keys-prefix '/openshift.io/oauth/accesstokens' \
+    --lease-duration 86400s
+$ oadm migrate etcd-ttl <certificate_details> <endpoint> \
+    --ttl-keys-prefix '/openshift.io/oauth/authorizetokens' \
+    --lease-duration 500s
+----
+
+. +++<b>Reconfigure the master:</b>+++
+
+.. After the migration is complete, the
+xref:../install_config/master_node_configuration.adoc#master-configuration-files[master
+configuration file] (the *_/etc/origin/master/master-config.yaml_* file by
+default), to be updated so the master daemons can use the new storage back end:
++
+[source,yaml]
+----
+kubernetesMasterConfig:
+  apiServerArguments:
+    storage-backend:
+    - etcd3
+    storage-media-type:
+    - application/vnd.kubernetes.protobuf
+----
+
+.. Restart your services; for single master clusters, run:
++
+----
+# systemctl start atomic-openshift-master
+----
++
+For multiple master clusters, run the following on all masters:
++
+----
+# systemctl start atomic-openshift-master-api \
+    atomic-openshift-master-controllers
+----
+
+[[etcd-data-migration-recovering]]
+== Recovering from Migration Issues
+
+If you discover problems after the migration has completed, you may wish to restore
+from a backup:
+
+. Stop the master services:
++
+----
+# systemctl stop atomic-openshift-master \
+    atomic-openshift-master-api \
+    atomic-openshift-master-controllers
+----
+
+. Remove the `storage-backend` and `storage-media-type` keys from from
+`kubernetesMasterConfig.apiServerArguments` section in the master configuration
+file on each master:
++
+[source,yaml]
+----
+kubernetesMasterConfig:
+  apiServerArguments:
+   ...
+----
+
+. Restore from backups that were taken prior to the migration, located in
```

Comment:
Should we suggest doing a backup before migrating or is this an automatic process?

Reply:
@sdodson Is the recover procedure assuming they ran the automated migration?

Reply:
The backup is run automatically before the migration.

Reply:
Added a link how to manually backup etcd to the manual steps.